### ISSUE TITLE
Motion dmd loss does not have gradient

### ISSUE THREAD

**Issue by xliuwave** [Sphere-AI-Lab/diagdistill#2]
Hi, Thank you very much for the code. It looks to me that the motion dmd loss does not have any gradient.

In this line, the grad_motion is computed inside torch.no_grad():

https://github.com/Sphere-AI-Lab/diagdistill/blob/4ea418d20ac18a9593eedc5e8df8ccbc50fa51a0/model/dmd.py#L204

After that, when you compute the weights for the motion loss in this line. It returns detached weights.

https://github.com/Sphere-AI-Lab/diagdistill/blob/4ea418d20ac18a9593eedc5e8df8ccbc50fa51a0/model/dmd.py#L232

Therefore, in the motion loss, nothing has gradient.

https://github.com/Sphere-AI-Lab/diagdistill/blob/4ea418d20ac18a9593eedc5e8df8ccbc50fa51a0/model/dmd.py#L246

Can you please look into this issue?

**Comment by Brandon-Liu-Jx** [comment]
> Hi, Thank you very much for the code. It looks to me that the motion dmd loss does not have any gradient.
> 
> In this line, the grad_motion is computed inside torch.no_grad():
> 
> [diagdistill/model/dmd.py](https://github.com/Sphere-AI-Lab/diagdistill/blob/4ea418d20ac18a9593eedc5e8df8ccbc50fa51a0/model/dmd.py#L204)
> 
> Line 204 in [4ea418d](/Sphere-AI-Lab/diagdistill/commit/4ea418d20ac18a9593eedc5e8df8ccbc50fa51a0)
> 
>  grad, grad_motion, dmd_log_dict = self._compute_kl_grad( 
> After that, when you compute the weights for the motion loss in this line. It returns detached weights.
> 
> [diagdistill/model/dmd.py](https://github.com/Sphere-AI-Lab/diagdistill/blob/4ea418d20ac18a9593eedc5e8df8ccbc50fa51a0/model/dmd.py#L232)
> 
> Line 232 in [4ea418d](/Sphere-AI-Lab/diagdistill/commit/4ea418d20ac18a9593eedc5e8df8ccbc50fa51a0)
> 
>  return (1.0 + cum_error / denom).detach() 
> Therefore, in the motion loss, nothing has gradient.
> 
> [diagdistill/model/dmd.py](https://github.com/Sphere-AI-Lab/diagdistill/blob/4ea418d20ac18a9593eedc5e8df8ccbc50fa51a0/model/dmd.py#L246)
> 
> Line 246 in [4ea418d](/Sphere-AI-Lab/diagdistill/commit/4ea418d20ac18a9593eedc5e8df8ccbc50fa51a0)
> 
>  weighted_squared_error = squared_error * hybrid_weights.view( 
> Can you please look into this issue?

Hi, thanks for your interest in our work! We have uploaded the latest version of our codebase, feel free to check this!
